raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -194,14 +196,52 @@ private void completeCurrentBatch() {
         MemoryRecords data = currentBatch.build();
         completed.add(new CompletedBatch<>(
             currentBatch.baseOffset(),
-            currentBatch.records(),
+            Optional.of(currentBatch.records()),
             data,
             memoryPool,
             currentBatch.initialBuffer()
         ));
         currentBatch = null;
     }
 
+    public void appendLeaderChangeMessage(LeaderChangeMessage leaderChangeMessage, long currentTimeMs) {
+        appendLock.lock();
+        try {
+            forceDrain();
+            ByteBuffer buffer = memoryPool.tryAllocate(256);
+            if (buffer != null) {
+                MemoryRecords data = MemoryRecords.withLeaderChangeMessage(
+                        this.nextOffset, 
+                        currentTimeMs, 
+                        this.epoch, 
+                        buffer, 
+                        leaderChangeMessage);
+                completed.add(new CompletedBatch<>(
+                    nextOffset,
+                    Optional.empty(),
+                    data,
+                    memoryPool,
+                    buffer
+                ));
+                nextOffset += 1;
+            } else {
+                throw new IllegalStateException("Could not allocate buffer for the control batch.");
```

Comment:
nit: perhaps we can be more explicit and say "leader change record" instead of "control batch"